### PR TITLE
Init sidepanel only when loading

### PR DESCRIPTION
**Pull request by gerhardol**
Follow up to #9864 

## Proposed changes

* Previously, the side panel was initiated from Browse (UICommands_PostRepositoryChanged is almost always invoked by Browse changes regardless or not if the repo is changed),
then again after loading the grid if a filter was applied or changed.
(normally the loaded branches etc were reused, so no additional Git call just making this confusing and some additional loading).
The second part is needed if the change was initiated from RevGrid itself (normally when changing filters etc).

The side panel init is now only started when loading the grid.
Update is forced if the repo changes (set for instance for F5 refreshes).

Some parts separated to #9896 and #9897

## Test methodology 

manual

## Merge strategy

Rebase merge 

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: GitUI/BranchTreePanel/RepoObjectsTree.BranchTree.cs [hunk 1]
```diff
@@ -39,12 +39,6 @@ protected override void OnAttached()
                 IsFiltering.Value = false;
             }
 
-            protected override Task PostRepositoryChangedAsync(GitUIEventArgs e)
-            {
-                IsFiltering.Value = false;
```
Comment (RussKie) [inline GitUI/BranchTreePanel/RepoObjectsTree.BranchTree.cs:44]: How is `IsFiltering` used now?

Comment (gerhardol) [reply GitUI/BranchTreePanel/RepoObjectsTree.BranchTree.cs:44]: Not changed at all, it is just that the panel is updated only once.
(Maybe the property can be removed?)

Comment (gerhardol) [reply GitUI/BranchTreePanel/RepoObjectsTree.BranchTree.cs:44]: `IsFiltering` is required but `_isCurrentlyFiltering` can be removed.
However, when testing this I get a race condition causing _isCurrentlyFiltering to be set before IsFiltering so if the refresh is initiated from the grid (filter only) or Browse (could be module), the result may differ.
I changed this to only fully refresh the side panel on  actual module changes, so refreshing from grid or browse is the same. 

Cannot see a need to force the refresh.
If so this can be handled in the Loading event.

Comment (gerhardol) [reply GitUI/BranchTreePanel/RepoObjectsTree.BranchTree.cs:44]: The loaded structure must be invalidated when changed (like a branch is added) so the latest commit will need to be adjusted.

File: GitUI/UserControls/RevisionGrid/RevisionGridControl.cs [hunk 2]
```diff
@@ -862,7 +860,7 @@ private void ShowLoading(bool showSpinner = true)
         ///  Queries git for the new set of revisions and refreshes the grid.
         /// </summary>
         /// <exception cref="AggregateException"></exception>
-        public void PerformRefreshRevisions(Func<RefsFilter, IReadOnlyList<IGitRef>> getRefs = null)
+        public void PerformRefreshRevisions(Func<RefsFilter, IReadOnlyList<IGitRef>> getRefs = null, bool updateSidepanel = false)
```
Comment (RussKie) [inline GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:863]: This name implies this control needs to know about some other unrelated control, which is undesirable.
```suggestion
        public void PerformRefreshRevisions(Func<RefsFilter, IReadOnlyList<IGitRef>> getRefs = null, bool forceRefresh = false)
```

Also please add the xmldoc explaining the parameter.

File: Plugins/GitUIPluginInterfaces/GitUIEventArgs.cs [hunk 3]
```diff
@@ -9,12 +9,13 @@ public class GitUIEventArgs : CancelEventArgs
     {
         private readonly IFilteredGitRefsProvider _getRefs;
 
-        public GitUIEventArgs(IWin32Window? ownerForm, IGitUICommands gitUICommands, Lazy<IReadOnlyList<IGitRef>> getRefs)
+        public GitUIEventArgs(IWin32Window? ownerForm, IGitUICommands gitUICommands, Lazy<IReadOnlyList<IGitRef>> getRefs, bool isForceUpdate)
```
Comment (RussKie) [inline Plugins/GitUIPluginInterfaces/GitUIEventArgs.cs:12]:
```suggestion
        public GitUIEventArgs(IWin32Window? ownerForm, IGitUICommands gitUICommands, Lazy<IReadOnlyList<IGitRef>> getRefs, bool forceRefresh)
```

File: Plugins/GitUIPluginInterfaces/GitUIEventArgs.cs [hunk 4]
```diff
@@ -32,5 +34,6 @@ public GitUIEventArgs(IWin32Window? ownerForm, IGitUICommands gitUICommands)
         public IGitModule GitModule => GitUICommands.GitModule;
 
         public IReadOnlyList<IGitRef> GetRefs(RefsFilter filter) => _getRefs.GetRefs(filter);
+        public bool IsForceUpdate { get; init; }
```
Comment (RussKie) [inline Plugins/GitUIPluginInterfaces/GitUIEventArgs.cs:37]:
```suggestion
        public bool ForceRefresh { get; init; }
```

File: GitUI/BranchTreePanel/RepoObjectsTree.cs [hunk 5]
```diff
@@ -252,11 +252,11 @@ public void Initialize(IAheadBehindDataProvider? aheadBehindDataProvider, Action
         ///  <see langword="true"/>, if the data is being filtered; otherwise <see langword="false"/>.
         /// </param>
         /// <param name="getRefs">Function to get refs.</param>
-        public void Refresh(bool isFiltering, Func<RefsFilter, IReadOnlyList<IGitRef>> getRefs)
+        public void Refresh(bool isFiltering, bool isForceUpdate, Func<RefsFilter, IReadOnlyList<IGitRef>> getRefs)
```
Comment (RussKie) [inline GitUI/BranchTreePanel/RepoObjectsTree.cs:255]:
```suggestion
        public void Refresh(bool isFiltering, bool forceRefresh, Func<RefsFilter, IReadOnlyList<IGitRef>> getRefs)
```

File: GitUI/BranchTreePanel/RepoObjectsTree.Nodes.cs [hunk 6]
```diff
@@ -124,12 +124,8 @@ private abstract class Tree : IDisposable
             private readonly CancellationTokenSequence _reloadCancellationTokenSequence = new();
             private bool _firstReloadNodesSinceModuleChanged = true;
 
-            // A flag to indicate whether the data is currently being filtered or not.
-            // This helps to reduce unnecessary treeview rebinds.
-            private bool _isCurrentlyFiltering;
-
             // A flag to indicate whether the data is being filtered (e.g. Show Current Branch Only).
-            private protected static AsyncLocal<bool> IsFiltering = new();
+            private protected AsyncLocal<bool> IsFiltering = new();
```
Comment (RussKie) [inline GitUI/BranchTreePanel/RepoObjectsTree.Nodes.cs:128]: Why did you remove the `static` modifier? Do we expect more than one instance of the left panel now?

Comment (gerhardol) [reply GitUI/BranchTreePanel/RepoObjectsTree.Nodes.cs:128]: This replaces the `_isCurrentlyFiltering` flag, it would have been cumbersome to keep two flags for the same info.

File: Plugins/GitUIPluginInterfaces/GitUIEventArgs.cs (hunk 3 above)
Comment (gerhardol) [reply Plugins/GitUIPluginInterfaces/GitUIEventArgs.cs:12]: will change this (and other comments)
plan to move this to a separate class as well, this is an odd fit here.

File: GitUI/BranchTreePanel/RepoObjectsTree.cs (hunk 5 above)
Comment (gerhardol) [reply GitUI/BranchTreePanel/RepoObjectsTree.cs]: done

File: Plugins/GitUIPluginInterfaces/GitUIEventArgs.cs (hunk 4 above)
Comment (gerhardol) [reply Plugins/GitUIPluginInterfaces/GitUIEventArgs.cs]: done

File: Plugins/GitUIPluginInterfaces/GitUIEventArgs.cs (hunk 3 above)
Comment (gerhardol) [reply Plugins/GitUIPluginInterfaces/GitUIEventArgs.cs]: done

File: GitUI/UserControls/RevisionGrid/RevisionGridControl.cs (hunk 2 above)
Comment (gerhardol) [reply GitUI/UserControls/RevisionGrid/RevisionGridControl.cs]: done

File: GitUI/CommandsDialogs/FormBrowse.cs [hunk 7]
```diff
@@ -736,7 +744,7 @@ private void RegisterPlugins()
                         if (plugin.Execute(new GitUIEventArgs(this, UICommands)))
                         {
                             _gitStatusMonitor.InvalidateGitWorkingDirectoryStatus();
-                            UICommands.RepoChangedNotifier.Notify();
+                            RefreshRevisions();
```
Comment (mstv) [inline GitUI/CommandsDialogs/FormBrowse.cs:747]: A drawback of this might be that a future subscriber of `UICommands.RepoChangedNotifier` might not get updated.

**Review by mstv — COMMENTED** [review]
Have not noticed side-effects so far.

File: GitUI/CommandsDialogs/FormBrowse.cs (hunk 7 above)
Comment (gerhardol) [reply GitUI/CommandsDialogs/FormBrowse.cs:747]: If the repo is changed, they will.
If a notification is required for other reasons, there should be a separate notification (like the existing GridLoading).
The intention of this was to make the handling possible to understand.

File: GitUI/UserControls/RevisionGrid/RevisionGridControl.cs [hunk 8]
```diff
@@ -62,7 +62,7 @@ public sealed partial class RevisionGridControl : GitModuleControl, IScriptHostC
     {
         public event EventHandler<DoubleClickRevisionEventArgs>? DoubleClickRevision;
         public event EventHandler<FilterChangedEventArgs>? FilterChanged;
-        public event EventHandler<GitUIEventArgs>? RevisionGraphLoaded;
+        public event EventHandler<GridLoadingEventArgs>? RevisionGraphLoading;
```
Comment (RussKie) [inline GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:65]: I suggest to retain the event to indicate the data is loaded as well. It may be useful for plugins.

```suggestion
        /// <summary>
        ///  Occurs whenever the revision graph has been populated with the data.
        /// </summary>
        public event EventHandler<DataLoadEventArgs>? DataLoaded;

        /// <summary>
        ///  Occurs whenever the revision graph has started loading the data.
        /// </summary>
        public event EventHandler<DataLoadEventArgs>? DataLoading;
```

Comment (gerhardol) [reply GitUI/UserControls/RevisionGrid/RevisionGridControl.cs:65]: I fam a little hesitant to add features without a direct need, but can add this.

File: GitUI/UserControls/RevisionGrid/GridLoadingEventArgs.cs [hunk 9]
```diff
@@ -0,0 +1,18 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Windows.Forms;
+using GitUIPluginInterfaces;
+
+namespace GitUI.UserControls.RevisionGrid
+{
+    public class GridLoadingEventArgs : GitUIEventArgs
+    {
+        public GridLoadingEventArgs(IWin32Window? ownerForm, IGitUICommands gitUICommands, Lazy<IReadOnlyList<IGitRef>> getRefs, bool forceRefresh)
```
Comment (RussKie) [inline GitUI/UserControls/RevisionGrid/GridLoadingEventArgs.cs:10]:
```suggestion
        public GridLoadEventArgs(IWin32Window? ownerForm, IGitUICommands gitUICommands, Lazy<IReadOnlyList<IGitRef>> getRefs, bool forceRefresh)
```

**Review by RussKie — COMMENTED** [review]
👍

File: GitUI/UserControls/RevisionGrid/GridLoadingEventArgs.cs (hunk 9 above)
Comment (gerhardol) [reply GitUI/UserControls/RevisionGrid/GridLoadingEventArgs.cs]: done

**Review by mstv — APPROVED** [review]

**Comment by gerhardol** [comment]
Rebased and squashed without changes, rebase merge tomorrow
